cpp/ql/src/semmle/code/cpp/ir/internal/CppType.qll

```diff
@@ -0,0 +1,500 @@
+private import cpp
+private import semmle.code.cpp.Print
+private import semmle.code.cpp.ir.implementation.IRType
+private import semmle.code.cpp.ir.implementation.raw.internal.IRConstruction as IRConstruction
+
+/**
+ * Works around an extractor bug where a function reference gets a size of one byte.
+ */
+private int getTypeSizeWorkaround(Type type) {
+  exists(Type unspecifiedType |
+    unspecifiedType = type.getUnspecifiedType() and
+    (
+      unspecifiedType instanceof FunctionReferenceType and
+      result = any(NullPointerType t).getSize()
+      or
+      exists(PointerToMemberType ptmType |
+        ptmType = unspecifiedType and
+        (
+          if ptmType.getBaseType().getUnspecifiedType() instanceof RoutineType
+          then result = any(NullPointerType t).getSize() * 2
+          else result = any(NullPointerType t).getSize()
+        )
+      )
+    )
+  )
+}
+
+private int getTypeSize(Type type) {
+  if exists(getTypeSizeWorkaround(type))
+  then result = getTypeSizeWorkaround(type)
+  else result = type.getSize()
```

Comment:
Done.
I would have expected two different pointer types, each with a different size.

Reply:
Cool. I've removed that `max()`. I've left the `max()` used in `getPointerSize()`, because I don't know of a way to tell the pointer size for a particular function. We'll just always assume the largest one.